### PR TITLE
Sign extension issue

### PR DESCRIPTION
**Pull request by AdrianCX**
Function from: https://doc.rust-lang.org/beta/src/core/up/stdarch/crates/core_arch/src/x86/sse2.rs.html#1377-1379

```
pub unsafe fn _mm_extract_epi16(a: __m128i, imm8: i32) -> i32 {
    simd_extract::<_, i16>(a.as_i16x8(), (imm8 & 7) as u32) as i32
}
```

Extracts as i16 and then converts i16 to i32... which does sign extension - instead of getting 0xffff we get 0xffffffff

This is contrary to intel documentation:
https://software.intel.com/sites/landingpage/IntrinsicsGuide/#text=_mm_extract_epi16&expand=2426

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -29,3 +29,4 @@ nightly = [] # Used only for testing
 [dev-dependencies]
 rand = "0.6"
 paste = "0.1"
+lazy_static = "1"
```
Comment (jethrogb) [inline Cargo.toml]: What is this?

File: src/decode/avx2.rs [hunk 2]
```diff
@@ -132,8 +132,11 @@ unsafe fn decode_avx2(input: __m256i) -> (__m256i, u32, u32) {
     let mask0 = _mm_cmpestrm(valid_nonws_set, 14, lane0, 16, CMP_FLAGS);
     let mask1 = _mm_cmpestrm(valid_nonws_set, 14, lane1, 16, CMP_FLAGS);
     // Combine bitmasks into integer value
```
Comment (jethrogb) [inline src/decode/avx2.rs]: This comment belongs with the code block below

File: src/decode/mod.rs [hunk 3]
```diff
@@ -131,22 +131,24 @@ fn decode64<D: Decoder, P: Packer>(input: &[u8], decoder: D, packer: P) -> Resul
...
+                    trailer_length = match (iter.next(), iter.next()) {
+                        (None, _) => Some(1),
+                        (Some((_, b'=')), None) => Some(2),
+                        (Some((_, b'=')), Some((i, _))) | (Some((i, _)), _) => {
+                            return Err(Error::InvalidCharacter(rest_start + i))
+                        }
+                    };
+                } else {
+                    return Err(Error::InvalidCharacter(chunk_start + idx));
+                }
```
Comment (jethrogb) [inline src/decode/mod.rs]: Please explain the need for this change

**Comment by jethrogb** [comment]
Please add some tests that cover the changes you made

File: Cargo.toml (hunk 1 above)
Comment (AdrianCX) [reply Cargo.toml]: removed

File: src/decode/mod.rs (hunk 3 above)
Comment (AdrianCX) [reply src/decode/mod.rs]: Reverted - I thought this was an issue but it turns out not.

File: src/decode/avx2.rs [hunk 4]
```diff
@@ -131,9 +131,12 @@ unsafe fn decode_avx2(input: __m256i) -> (__m256i, u32, u32) {
...
     let mask0 = _mm_cmpestrm(valid_nonws_set, 14, lane0, 16, CMP_FLAGS);
     let mask1 = _mm_cmpestrm(valid_nonws_set, 14, lane1, 16, CMP_FLAGS);
+
+    let first = _mm_extract_epi16(mask0, 0) as u16;
+    let second = _mm_extract_epi16(mask1, 0) as u16;
+    
     // Combine bitmasks into integer value
-    let valid_mask =
-        _mm_extract_epi16(mask0, 0) as u32 | ((_mm_extract_epi16(mask1, 0) as u32) << 16);
+    let valid_mask = first as u32 + ((second as u32) << 16);
```
Comment (jethrogb) [inline src/decode/avx2.rs:138]: Sorry this is not what I meant before. I was referring to the whitespace under the comment and not before. Final should look this:
```
    let mask0 = _mm_cmpestrm(valid_nonws_set, 14, lane0, 16, CMP_FLAGS);
    let mask1 = _mm_cmpestrm(valid_nonws_set, 14, lane1, 16, CMP_FLAGS);

    // Combine bitmasks into integer value
    let first = _mm_extract_epi16(mask0, 0) as u16;
    let second = _mm_extract_epi16(mask1, 0) as u16;
    let valid_mask = first as u32 + ((second as u32) << 16); 
```

**Review by jethrogb — APPROVED** [review]
Please bump patch version in Cargo.toml

**Pull request merged**